elasticdl/python/worker/task_data_service.py

```diff
@@ -114,12 +114,8 @@ def _gen(self):
         """
         while True:
             task = self._data_shard_service.get_task()
-            if not task.shard.name:
+            if task.type != elasticdl_pb2.TRAINING:
                 break
-            with self._lock:
-                if task.type == elasticdl_pb2.TRAIN_END_CALLBACK:
-                    self._pending_train_end_callback_task = task
-                    break
```

Comment:
Here, we don't need to check whether the task is `TRAIN_END_CALLBACK`. Because the task must be `WAIT` after Training if there are also training tasks is not completed. The task will be `TRAIN_END_CALLBACK` only after all training tasks are completed. 
https://github.com/sql-machine-learning/elasticdl/blob/d7d081b43ae05525d83d73aa6d430ececfcf5ba5/elasticdl/python/master/servicer.py#L105-L119

After the training loop is completed, the worker will get the `TRAIN_END_CALLBACK` task in codes:

https://github.com/sql-machine-learning/elasticdl/blob/d7d081b43ae05525d83d73aa6d430ececfcf5ba5/elasticdl/python/worker/task_data_service.py#L80-L94